Intake Changes

Yumnah2 opened the pull request.

Yumnah2 committed:
Added intake sensor debouncer and feeder sensor

Yumnah2 committed:
Merge branch 'main' of https://github.com/Yumnah2/Crescendo2024

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -30,52 +30,55 @@ public void initialize() {
-		// 		rumbledIntakeRight = true;
-		// 	}
-		// }
+		if (intakeSubsystem.intakeFrontSeesNote()) {
```
The intake sensors were taken off, so we should leave this commented out. The index and feeder sensors are still there though.

Yumnah2 committed:
commented removed sensors, debounced feeder sensor

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -46,9 +46,15 @@ public class IntakeSubsystem extends SubsystemBase {
 	// Sensors
 	private final SparkLimitSwitch indexSensor;
-	private final DigitalInput feederSensor;
-
+	private final SparkLimitSwitch feederSensor;
```
I thought feeder now physically has two sensors (the existing IR reflective wired to digital input, and the new beam break wires to the SparkFlex controller). If that's true, I'd recommend reading both & ORing the result so either triggering will cause feeder to stop.

Yumnah2 replied:
currently is set to OR (might have to change later if the first sensor prevents the note from going all the way to the flywheels)

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -81,18 +87,26 @@ public IntakeSubsystem() {
 		intakeFrontSensor.enableLimitSwitch(false);
 		intakeLeftSensor.enableLimitSwitch(false);
 		intakeRightSensor.enableLimitSwitch(false);
+		feederSensor.enableLimitSwitch(false);
```
I believe these are already set in the configure motor, so can be deleted.

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -219,7 +233,7 @@ public boolean indexSensorHasNote() {
 
 	public boolean feederSensorHasNote() {
-		return !feederSensor.get() && !getSensorOverride();
+		return feederSensor.isPressed() && !getSensorOverride();
```
Should use `debouncedFeederSensor()` instead of directly using `feederSensor`

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -81,18 +87,26 @@ public IntakeSubsystem() {
+		intakeRightSensorDebouncer = new Debouncer(0.1);
+		intakeLeftSensorDebouncer = new Debouncer(0.1);
+
+		feederSensorDebouncer = new Debouncer(0.1);
```
I thought one issue was the sensor briefly reading 'false' even when a note was held. If that's true, then per https://docs.wpilib.org/en/stable/docs/software/advanced-controls/filters/debouncer.html this should pass in a 2nd argument to specify that both the rising (note detected) & falling (note no longer detected) cases should be debounced.

Yumnah2 replied:
oops i missed that

Yumnah2 committed:
addressing comments

added IR feeder sensor, changed debouncers to Both, feeder stops if either sensor triggers

Yumnah2 committed:
fixed some bugs

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -218,8 +229,13 @@ public boolean indexSensorHasNote() {
 	}
 
+	public boolean debouncedFeederSensor() {
+		return feederSensorDebouncer.calculate(feederSensor.isPressed());
```
To simplify the rest of the code & de-bounce both feeder sensors, try this:
```suggestion
		boolean feederSensorSignal = feederSensorDebouncer.calculate(feederSensor.isPressed());
		boolean feederSensorIRSignal = feederSensorIRDebouncer.calculate(feederSensorIR.get());
		return feederSensorSignal  || feederSensorIRSignal ;
```

jamesdooley4 replied:
And remove the use of `feederSensorIR` from `feederSensorHasNote()`

kirbt committed:
debouncer changes

kirbt committed:
unstati

kirbt committed:
yohohoho

merry christmas jameson

jamesdooley4 commented:
@Yumnah2 - closing this PR, as it was merged into https://github.com/robototes/Crescendo2024/pull/114 and tested & updated there.
